Only set sample clips for valid files

michaelgregorius opened the pull request:
Check if a non-empty buffer was loaded and only set the sample clip if that's the case.

Fix code formatting.

Fixes #6287.

sakertooth commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -183,16 +182,22 @@ void SampleClipView::mouseDoubleClickEvent( QMouseEvent * )
+	if (af == m_clip->m_sample.sampleFile())
+	{
+		// Don't reload the same file again. Just reset the size.
+		m_clip->changeLengthToSampleLength();
```
Was this refactoring necessary?

michaelgregorius replied:
It's more or less an application of the [boyscout rule](https://deviq.com/principles/boy-scout-rule). The view contained logic/actions that's part of the core/"business" and therefore I moved it to where it belongs.

sakertooth replied:
Since its low hanging fruit, it's not a big deal at all, but I would add some details about it in the commit description (something along the lines of also move around `setModified` logic to the correct spot) since this PR is doing more than what it said it would on the tin.

michaelgregorius replied:
It's already mentioned in the commit message of commit 14ce3cc0.

sakertooth commented on src/gui/clips/SampleClipView.cpp:
```diff
@@ -183,16 +182,22 @@ void SampleClipView::mouseDoubleClickEvent( QMouseEvent * )
+		// Don't reload the same file again. Just reset the size.
+		m_clip->changeLengthToSampleLength();
 	}
 	else
```
You could combine this like so:
```suggestion
	else if (auto sampleBuffer = SampleLoader::createBufferFromFile(af); sampleBuffer != SampleBuffer::emptyBuffer()) 
```

michaelgregorius replied:
Thanks for the suggestion but I find it more readable and quicker to comprehend as is.

sakertooth commented on src/core/SampleClip.cpp:
```diff
@@ -129,6 +134,8 @@ void SampleClip::setSampleBuffer(std::shared_ptr<const SampleBuffer> sb)
 
 	emit sampleChanged();
+
+	Engine::getSong()->setModified();
```
Is this in scope?

michaelgregorius replied:
Same as above with regards to the boyscout rule. Setting the song to modified because something has changed in the core/"business" is not in the responsibility of the presentation/GUI layer. With the changes it will work even if the GUI was switched completely of if there was no GUI and some other entity would listen to song changes.

sakertooth commented on src/core/SampleClip.cpp:
```diff
@@ -210,6 +217,8 @@ void SampleClip::setIsPlaying(bool isPlaying)
 {
 	emit sampleChanged();
+
+	Engine::getSong()->setModified();
```
Same here?

michaelgregorius replied:
Same answer as above. :wink:

sakertooth reviewed (commented):
The comments are a bit much. I feel like the code is simple enough that it can be understood very well without them.

Veratil commented on src/core/SampleClip.cpp:
```diff
@@ -112,6 +112,11 @@ void SampleClip::changeLength( const TimePos & _length )
 
+void SampleClip::changeLengthToSampleLength()
+{
+	int length = static_cast<int>(m_sample.sampleSize() / Engine::framesPerTick());
```
Do we need the static_cast if we're already assigning to int? Or is this to suppress a warning?

michaelgregorius replied:
You're right. We don't need it. It was "inherited" from the existing code and I moved it without putting too much thought into it. Removed with commit ef681e1b79a.

sakertooth commented:
Do we really need all of these comments? The code is very simple to understand without them.

michaelgregorius commented:
> Do we really need all of these comments? The code is very simple to understand without them.

Most of the comments have been kept and slightly adjusted from the previous code. With commit 4d5bf7c6078 I have removed them. I have also introduced `SampleClip::hasSampleFileLoaded` to make the check more readable so that the comment can also be removed there. Before it was not directly obvious that the string comparison checks if the filename corresponds to the file that's loaded in the clip.
